File: 3189-find-champion-ii/3189-find-champion-ii.cpp
class Solution {
public:
    int findChampion(int n, vector<vector<int>>& edges) {
        unordered_map<int, int> indegree;
        
        for(const auto& edge : edges) {
            indegree[edge[1]]++;
        }
        
        vector<int> good;
        for(int i = 0; i < n; i++) {
            if (indegree.find(i) == indegree.end()) {
                good.push_back(i);
            }
        }
        
        if (good.size() == 1) {
            return good[0];
        }
        
        return -1;
    }
};

File: 3189-find-champion-ii/3189-find-champion-ii.java
import java.util.*;

class Solution {
    public int findChampion(int n, int[][] edges) {
        Map<Integer, Integer> indegree = new HashMap<>();
        
        for (int[] edge : edges) {
            int endNode = edge[1];
            indegree.put(endNode, indegree.getOrDefault(endNode, 0) + 1);
        }
        
        List<Integer> good = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (!indegree.containsKey(i)) {
                good.add(i);
            }
        }
        
        if (good.size() == 1) {
            return good.get(0);
        }
        
        return -1;
    }
}

public class Main {
    public static void main(String[] args) {
        int n = 5;
        int[][] edges = {{0, 1}, {1, 2}, {2, 3}, {3, 4}, {4, 0}};
        int ret = new Solution().findChampion(n, edges);
        System.out.println(ret); // Output: -1
    }
}
